lib/requestwrapper.ts

```diff
@@ -20,23 +20,276 @@ import FormData = require('form-data');
 import https = require('https');
 import querystring = require('querystring');
 import { PassThrough as readableStream } from 'stream';
+import {HeaderOptions} from "./base_service";
 import { buildRequestFileObject, getMissingParams, isEmptyObject, isFileParam } from './helper';
 
 // tslint:disable-next-line:no-var-requires
 const pkg = require('../package.json');
 const isBrowser = typeof window === 'object';
 const globalTransactionId = 'x-global-transaction-id';
 
-// axios sets the default Content-Type for `post`, `put`, and `patch` operations
-// to 'application/x-www-form-urlencoded'. This causes problems, so overriding the
-// defaults here
-['post', 'put', 'patch'].forEach(method => {
-  axios.defaults.headers[method]['Content-Type'] = 'application/json';
-});
+const allowedAxiosConfig = ['transformRequest', 'transformResponse', 'paramsSerializer', 'paramsSerializer', 'timeout', 'withCredentials', 'adapter', 'responseType', 'responseEncoding', 'xsrfCookieName', 'xsrfHeaderName', 'onUploadProgress', 'onDownloadProgress', 'maxContentLength', 'validateStatus', 'maxRedirects', 'socketPath', 'httpAgent', 'httpsAgent', 'proxy', 'cancelToken'];
+
+export class RequestWrapper {
+  private axiosInstance;
+
+  constructor(axiosOptions) {
+    // override several axios defaults
+    // axios sets the default Content-Type for `post`, `put`, and `patch` operations
+    // to 'application/x-www-form-urlencoded'. This causes problems, so overriding the
+    // defaults here
+    const axiosConfig = {
+      httpsAgent: new https.Agent({
+        rejectUnauthorized: axiosOptions.rejectUnauthorized
+      }),
+      maxContentLength: Infinity,
+      headers: {
+        post: {
+          'Content-Type':'application/json'
+        },
+        put: {
+          'Content-Type':'application/json'
+        },
+        patch: {
+          'Content-Type':'application/json'
+        },
+      }
+    };
+
+    // merge valid Axios Config into default.
+    extend(true, axiosConfig, allowedAxiosConfig.reduce((a,b)=> (a[b]=axiosOptions[b], a),{}));
```

Comment:
Would you alter this a little for readability? Either name `a` and `b` differently like `opts` and `key`, or just add a comment to describe what it's doing